lib/src/cancelable_operation.dart

```diff
@@ -497,9 +507,59 @@ class CancelableCompleter<T> {
 
     if (_inner != null) {
       _inner = null;
-      var onCancel = _onCancel;
-      cancelCompleter.complete(onCancel == null ? null : Future.sync(onCancel));
+      cancelCompleter.complete(_invokeCancelCallbacks());
     }
     return cancelCompleter.future;
   }
+
+  /// Invoke [_onCancel] and forward to other completers in [_cancelForwarders].
+  ///
+  /// Returns the same value as [_onCancel]. Legacy uses may return a value
+  /// despite the signature having `void` return.
+  Future<Object?> _invokeCancelCallbacks() async {
+    final FutureOr<Object?> toReturn = _onCancel?.call();
+    final isFuture = toReturn is Future;
+    final cancelFutures = <Future<Object?>>[
+      if (isFuture) toReturn,
+      ...?_cancelForwarders?.map(_forward).whereNotNull()
+    ];
+    final results = (isFuture && cancelFutures.length == 1)
+        ? [await toReturn]
+        : cancelFutures.isNotEmpty
+            ? await Future.wait(cancelFutures)
+            : const [];
+    return isFuture ? results.first : toReturn;
```

Comment:
LGTM. But just because I can't not optimize, consider:

```dart
  if (cancelFutures.isEmpty) return toReturn;
  if (cancelFutures.length == 1) {
     var result = await cancelFutures[0];
     return isFuture ? result : toReturn;
  }
  var results = await Future.wait(cancelFutures);
  return isFuture ? results[0] : toReturn;
```

(Probably not worth it.)